demmt/object_gf100_compute.c

```diff
@@ -0,0 +1,181 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "buffer.h"
+#include "config.h"
+#include "demmt.h"
+#include "log.h"
+#include "nvrm.h"
+#include "object.h"
+
+struct gf100_compute_data
+{
+	struct addr_n_buf code;
+	struct addr_n_buf tsc;
+	uint32_t linked_tsc;
+	struct addr_n_buf tic;
+	struct addr_n_buf cb;
+	int cb_pos;
+	struct addr_n_buf image[8];
+
+	struct rnndeccontext *texture_ctx;
+
+	struct mthd2addr *addresses;
+};
+
+static void destroy_gf100_compute_data(struct gpu_object *obj)
+{
+	struct gf100_compute_data *d = obj->class_data;
+	rnndec_freecontext(d->texture_ctx);
+	free(d->addresses);
+	free(d);
+}
+
+void decode_gf100_compute_init(struct gpu_object *obj)
+{
+	struct gf100_compute_data *d = obj->class_data = calloc(1, sizeof(struct gf100_compute_data));
+	d->texture_ctx = create_g80_texture_ctx(obj);
+	obj->class_data_destroy = destroy_gf100_compute_data;
+
+#define SZ 6
+	struct mthd2addr *tmp = d->addresses = calloc(SZ, sizeof(*d->addresses));
+	m2a_set1(tmp++, 0x155c, 0x1560, &d->tsc);
+	m2a_set1(tmp++, 0x1574, 0x1578, &d->tic);
+	m2a_set1(tmp++, 0x1608, 0x160c, &d->code);
+	m2a_set1(tmp++, 0x2384, 0x2388, &d->cb);
+	m2a_setN(tmp++, 0x2700, 0x2704, &d->image[0], 8, 0x20);
+	m2a_set1(tmp++, 0, 0, NULL);
+	assert(tmp - d->addresses == SZ);
+#undef SZ
+}
+
+void decode_gf100_compute_terse(struct gpu_object *obj, struct pushbuf_decode_state *pstate)
+{
+	struct gf100_compute_data *objdata = obj->class_data;
+
+	if (check_addresses_terse(pstate, objdata->addresses))
+	{ }
+}
+
+void decode_gf100_compute_verbose(struct gpu_object *obj, struct pushbuf_decode_state *pstate)
+{
+	int mthd = pstate->mthd;
+	uint32_t data = pstate->mthd_data;
+	struct gf100_compute_data *objdata = obj->class_data;
+
+	if (check_addresses_verbose(pstate, objdata->addresses))
+	{ }
+	else if (mthd == 0x1234)
+	{
+		objdata->linked_tsc = data;
+	}
+	else if (mthd == 0x238c) // CB_POS
+	{
+		mmt_debug("cb_pos: 0x%08x\n", data);
+		if (objdata->cb.gpu_mapping)
+		{
+			objdata->cb_pos = data;
+		}
+	}
+	else if (mthd == 0x2390) // CB_DATA[0]
+	{
+		mmt_debug("cb_data: 0x%08x\n", data);
+		if (objdata->cb.gpu_mapping)
+		{
+			gpu_mapping_register_write(objdata->cb.gpu_mapping,
+						   objdata->cb.address + objdata->cb_pos,
+						   4, &data);
+			objdata->cb_pos += 4;
+		}
+	}
+	else if (mthd == 0x03b4) // CP_START_ID
+	{
+		struct varinfo *var;
+
+		if (!isa_gf100)
+			isa_gf100 = ed_getisa("gf100");
+
+		var = varinfo_new(isa_gf100->vardata);
+		varinfo_set_variant(var, "gf100");
+
+		// blob has been observed to feed an unmapped address as the
```

Comment:
Please keep the logic similar to the other logic... first try the code.mapping. Then if (!m) { ... }. What I was asking you originally was whether the if (!m) could be removed entirely... meaning the whole block, not just doing it unconditionally.

Also while you're at it, make sure to deal with gpu_mapping_find returning NULL.